ada_feeding_perception/ada_feeding_perception/segment_all_items.py

```diff
@@ -0,0 +1,1279 @@
+        item_id: The item ID of the mask.
+        object_id: The object ID of the mask.
+        score: The confidence score outputed by the segmentation model for the mask.
+        mask: The pixel-wise mask detected.
+        image: The image the mask was detected on.
+        depth_img: The most recent depth image.
+        bbox: The bounding box from GroundingDINO.
+        """
+        # Calculate center of the bounding box and use it as the seed point for
+        # getting the connected component of the mask
+        center_x = (bbox[0] + bbox[2]) // 2
+        center_y = (bbox[1] + bbox[3]) // 2
+        center_x = int(center_x)
+        center_y = int(center_y)
+        seed_point = (center_x, center_y)
+
+        # Use the mask to get a connected component containing the seed point
+        cleaned_mask = get_connected_component(mask, seed_point)
+
+        # Use the cleaned mask to calculate the median depth over the mask
+        masked_depth = depth_img[cleaned_mask]
+        median_depth_mm = np.median(
+            masked_depth[
+                np.logical_and(
+                    masked_depth >= self.min_depth_mm, masked_depth <= self.max_depth_mm
+                )
+            ]
+        )
+        # If the depth is invalid, skip this mask and return None
+        if np.isnan(median_depth_mm):
+            self._node.get_logger().warn(
+                f"No depth points within [{self.min_depth_mm}, {self.max_depth_mm}] mm range "
+                f"for mask {item_id}. Skipping mask."
+            )
+            return None
+
+        # Convert the bounding box from a Python Tuple into the BoundingBox class
+        # from helpers.py to call the crop_image_mask_and_point
+        bbox_converted = BoundingBox(
+            int(bbox[0]), int(bbox[1]), int(bbox[2]), int(bbox[3])
+        )
+
+        # Crop the image and the mask
+        cropped_image, cropped_mask, _ = crop_image_mask_and_point(
+            image, cleaned_mask, seed_point, bbox_converted
+        )
+        cropped_depth, _, _ = crop_image_mask_and_point(
+            depth_img, cleaned_mask, seed_point, bbox_converted
+        )
+
+        # Convert the mask to an image
+        mask_img = np.where(cropped_mask, 255, 0).astype(np.uint8)
+
+        # Create the ROS mask message
+        mask_msg = Mask()
+        mask_msg.roi = RegionOfInterest(
+            x_offset=int(bbox[0]),
+            y_offset=int(bbox[1]),
+            height=int(bbox[3] - bbox[1]),
+            width=int(bbox[2] - bbox[0]),
+            do_rectify=False,
+        )
+        mask_msg.mask = cv2_image_to_ros_msg(mask_img, compress=True)
+        mask_msg.rgb_image = cv2_image_to_ros_msg(cropped_image, compress=True)
+        mask_msg.depth_image = cv2_image_to_ros_msg(cropped_depth, compress=False)
+        mask_msg.average_depth = median_depth_mm / 1000.0
+        mask_msg.item_id = item_id
+        mask_msg.object_id = object_id
+        mask_msg.confidence = float(score)
```

Comment:
Old code for `SegmentFromPoint` didn't require a type-cast to `float`, it's fine if this is needed just wanted to check since I noticed the discrepancy.

Reply:
Also, these small changes are precisely why its best to not copy-paste code (e.g., if casting to `float` improves the code here, we should ideally percolate it back to `SegmentFromPoint`). So this is one more of the functions that should be in a helper.

Another thing you can consider is making `SegmentAllItems` **inherit** from the `SegmentFromPoint` class, and only override some functions.